cub/cub/detail/fast_modulo_division.cuh

```diff
@@ -0,0 +1,221 @@
+
+template <typename T>
+using implicit_prom_t = decltype(+T{});
+
+template <typename T>
+using unsigned_implicit_prom_t = typename ::cuda::std::make_unsigned<implicit_prom_t<T>>::type;
+
+/***********************************************************************************************************************
+ * Extract higher bits after multiplication
+ **********************************************************************************************************************/
+
+template <typename T, typename R>
+_CCCL_NODISCARD _CCCL_HOST_DEVICE _CCCL_FORCEINLINE unsigned_implicit_prom_t<T>
+multiply_extract_higher_bits(T value, R multiplier)
+{
+  static_assert(::cuda::std::is_integral<T>::value && !::cuda::std::is_same<T, bool>::value, "unsupported type");
+  static_assert(::cuda::std::is_integral<R>::value && !::cuda::std::is_same<R, bool>::value, "unsupported type");
+  _CCCL_ASSERT(value >= 0, "value must be non-negative");
+  _CCCL_ASSERT(multiplier >= 0, "multiplier must be non-negative positive");
+  using unsigned_t           = unsigned_implicit_prom_t<T>;
+  using larger_t             = larger_unsigned_type_t<T>;
+  constexpr unsigned BitSize = sizeof(T) * CHAR_BIT;
+  // clang-format off
+  NV_IF_TARGET(
+    NV_IS_HOST,
+      (return static_cast<unsigned_t>((static_cast<larger_t>(value) * multiplier) >> BitSize);),
+    //NV_IS_DEVICE
+      (return (sizeof(T) == 8)
+        ? static_cast<unsigned_t>(__umul64hi(value, multiplier))
+        : static_cast<unsigned_t>((static_cast<larger_t>(value) * multiplier) >> BitSize);));
+  // clang-format on
+}
+
+/***********************************************************************************************************************
+ * Fast Modulo/Division based on Precomputation
+ **********************************************************************************************************************/
+
+template <typename T>
+struct fast_div_mod
+{
+  static_assert(::cuda::std::is_integral<T>::value && !::cuda::std::is_same<T, bool>::value, "unsupported type");
+
+  using unsigned_t = unsigned_implicit_prom_t<T>;
+  using prom_t     = implicit_prom_t<T>;
+  using larger_t   = larger_unsigned_type_t<T>;
+
+  struct result
+  {
+    prom_t quotient;
+    prom_t remainder;
+  };
+
+  fast_div_mod() = delete;
+
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE explicit fast_div_mod(T divisor) noexcept
+      : _divisor{static_cast<prom_t>(divisor)}
+  {
+    _CCCL_ASSERT(divisor > 0, "divisor must be positive");
+    auto udivisor = static_cast<unsigned_t>(divisor);
+    if (::cuda::std::has_single_bit(udivisor)) // power of two
+    {
+      _shift_right = ::cuda::std::bit_width(udivisor) - 1;
+      return;
+    }
+    else if (sizeof(T) == 8 && divisor == 3)
+    {
+      return;
+    }
+    constexpr int BitSize   = sizeof(T) * CHAR_BIT;
+    constexpr int BitOffset = BitSize / 16;
+    int num_bits            = ::cuda::std::bit_width(udivisor) + !::cuda::std::has_single_bit(udivisor);
```

Comment:
```suggestion
    const int num_bits            = ::cuda::std::bit_width(udivisor) + !::cuda::std::has_single_bit(udivisor);
```

Reply:
I don't see any value of adding `const` here honestly

Reply:
The value is for the reader to know that the variable is not going to change later in this function. For me, it reduces the head space required to understand what a function does.

But we can agree to disagree.